### PR TITLE
[Backport] [2.x] Update bundled JDK to JDK-21.0.1 (#10576)

### PR DESCRIPTION
**Pull request by reta**
Backport of https://github.com/opensearch-project/OpenSearch/pull/10576 to `2.x`

### CODE REVIEW

**Comment by reta** [comment]
@peternied @nknize @andrross @dblock bumping bundled JDK to `21.0.1` in `2.x` (starting from 2.12) if there is no objection, so we could:
 - let people benefit from Apache Lucene 9.8 Panama integrations (vector & FFM)
 - let our ecosystem benefit from Panama (SIMD & vector)

**Comment by peternied** [comment]
@reta I don't see any CI changes, shouldn't the `gradle check` be running on jdk 11, 17, 21 with this change?  Other than local building, how do we know this works broadly

**Comment by peterzhuamazon** [comment]
cc: @CEHENKLE @bbarani on this change for 2.x.

**Comment by reta** [comment]
> @reta I don't see any CI changes, shouldn't the `gradle check` be running on jdk 11, 17, 21 with this change? Other than local building, how do we know this works broadly

@peternied CI should pick JDK version from the `version.properties` (so no others changes are required), the runner agents should be running new images shortly

**Comment by reta** [comment]
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29308/

https://github.com/opensearch-project/OpenSearch/issues/9332
https://github.com/opensearch-project/OpenSearch/issues/10798

**Comment by reta** [comment]
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29313/

https://github.com/opensearch-project/OpenSearch/issues/10152
https://github.com/opensearch-project/OpenSearch/issues/10798
https://github.com/opensearch-project/OpenSearch/issues/10303

**Comment by reta** [comment]
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29320/

https://github.com/opensearch-project/OpenSearch/issues/7934
https://github.com/opensearch-project/OpenSearch/issues/6090
https://github.com/opensearch-project/OpenSearch/issues/9499

File: buildSrc/version.properties
Comment (peternied) [inline buildSrc/version.properties:5]: Does this mean the tests don't run against 17 anymore?

Comment (reta) [reply buildSrc/version.properties:5]: Correct, we run against bundled JDK version (21 in this case)

**Comment by reta** [comment]
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/29367/

https://github.com/opensearch-project/OpenSearch/issues/10798

**Review by peternied — APPROVED** [review]

File: buildSrc/version.properties
Comment (reta) [reply buildSrc/version.properties:5]: >  I was having trouble seeing if we validate against other jre (11,17, 21) or platforms (linux, win) - do we have that setup and I missed it?

We do but not for Gradle check

**Review by peternied — APPROVED** [review]
@reta I'm going to approve but as there might be downstream impact, please merge when you think it the best time

**Comment by reta** [comment]
> @reta I'm going to approve but as there might be downstream impact, please merge when you think it the best time

Thanks a lot @peternied , I am waiting for @nknize @andrross @dblock @CEHENKLE signoff before moving forwards, the impact of this change (going unnoticed) is understood :)

**Comment by dblock** [comment]
I don't know what the implications of this change are.

**Comment by reta** [comment]
> At a high level, this might cause Gradle compatibility issues across multiple repos so we need to start a campaign soon.

Thanks @bbarani, for Gradle  specifically - no issues and no changes, this is only runtime change (which is as important as Gradle): we would need to make sure that all plugins work under JDK-21 as a runtime. 

> I don't know what the implications of this change are.

Thanks @dblock , explained above

**Review by dblock — APPROVED** [review]
I don't have any strong opinions of when to merge this, your call @reta, PR approved. 

IMO downstream plugin pain is not solvable without the kind of decoupling introduced by extensions

**Review by andrross — APPROVED** [review]

**Review by heemin32 — APPROVED** [review]

**Comment by heemin32** [comment]
Wouldn't it be better to merge early than later if we are targeting 2.12?

**Comment by reta** [comment]
> Wouldn't it be better to merge early than later if we are targeting 2.12?

Sure, but this change needs awareness and alignment that it is OK to target 2.12

**Comment by bbarani** [comment]
@dblock @reta @andrross What are the next steps? We will start incorporating the build changes in 2.x line based on the outcome.

**Comment by reta** [comment]
Seeing no objections, merging JDK-21 into 2.x branch

**Pull request merged**
